.github/workflows/rust.yml

```diff
@@ -19,6 +19,8 @@ jobs:
       - uses: Swatinem/rust-cache@v2.2.0
       - name: Run cargo check
         run: cargo check
+      - name: Run cargo fmt check
+        run: cargo fmt --check
```

Comment:
Yes, that's good point. But pre commit hooks needs to be manually installed after clone, so it also might be forgotten.

Other options might be:
* adding make file with `make check` which will do check and formatting as well.
* adding `CONTRIBUTING` file with general guidelines, including formatting might help.

But having it in CI is 100% guarantee that it code would be formatted. Having formatted code in main makes sure that if anyone runs `cargo format` it won't produce huge diff originally from untouched files.

Reply:
Do you think somebody might be scared away by failed formatting job and won't contribute anymore?

Reply:
Yeah, I've definitely seen that become an issue - contributors will do some good feature work but then CI fails and it never gets over the line. (It's especially bad projects you also enforce `clippy` in CI). Maybe adding a `makefile` which runs `cargo fmt`, `cargo check --deny-warnings` and `cargo clippy` would help lower the burden. 

(And, tbc, I'm ok with adding this check to CI as well - just want to minimize the overhead on contributors)

Reply:
I have added make command with check and fmt, but no clippy for now, because we need fix some clippy warnings first. 

Also I've added helper message in case if fmt check fails, so it should be easier to fix the problem.

> contributors will do some good feature work but then CI fails and it never gets over the line

That's pity. But overall, what if there's a feature and has no tests, will we accept that? 

> (And, tbc, I'm ok with adding this check to CI as well - just want to minimize the overhead on contributors)

I understand and I also want to have opportunity to contribute with as little friction as possible. Balance is important. As I mentioned before, if we ignore formatting, then it will be more hassle for those devs who have habit of running fmt and clippy, they will see many unrelated errors and that might affect their perception of the project.

Reply:
Chatted with @citizen-stig offline. We think the combination of CI plus a single command you can run to ensure that CI will pass (`make check`) strikes a nice balance. Marking this conversation as resolved!